### PR TITLE
[iOS] Fix data race related to RCTNetworkTask.status

### PR DESCRIPTION
**Pull request by hakonk**
Fix entails using non-synthesized getter, such that underlying backing is an std::atomic<RCTNetworkTaskStatus>.

In the greater scheme of things, I believe `RCTNetworkTask` should be improved as it has several `nonatomic` properties that are read and written to on different threads. Thread safety of this class seems to have been addressed on a per property basis, judging from the employment of `std::mutex` elsewhere in the implementation.

## Summary:

This is an attempt at fixing https://github.com/facebook/react-native/issues/44687.

## Changelog:

[iOS][FIXED] - Fix data race related to access on `RCTNetworkTask.status`.

## Test Plan:

Added unit test in class `RCTNetworkTaskTests`.

### CODE REVIEW

File: packages/react-native/Libraries/Network/RCTNetworkTask.h [hunk 1]
```diff
@@ -35,13 +35,12 @@ typedef NS_ENUM(NSInteger, RCTNetworkTaskStatus) {
...
 
-@property (nonatomic, readonly) RCTNetworkTaskStatus status;
-
 - (instancetype)initWithRequest:(NSURLRequest *)request
                         handler:(id<RCTURLRequestHandler>)handler
                   callbackQueue:(dispatch_queue_t)callbackQueue NS_DESIGNATED_INITIALIZER;
 
 - (void)start;
 - (void)cancel;
+- (RCTNetworkTaskStatus)getNetworkTaskStatus;
```
Comment (javache) [inline packages/react-native/Libraries/Network/RCTNetworkTask.h:44]: This should just be `status`.

> Use “get” only for methods that return objects and values indirectly

https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/CodingGuidelines/Articles/NamingMethods.html

You could also keep this as an `@property` and implement a getter that uses the instance field, so the dot-syntax doesn't need to be changed.

File: packages/rn-tester/RNTesterPods.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -863,6 +867,7 @@
 			isa = XCBuildConfiguration;
 			buildSettings = {
 				ALWAYS_SEARCH_USER_PATHS = NO;
```
Comment (javache) [inline packages/rn-tester/RNTesterPods.xcodeproj/project.pbxproj:869]: Please revert changes to the build config.

**Review by javache — COMMENTED** [review]
Thanks for improving this!

File: packages/react-native/Libraries/Network/RCTNetworkTask.mm [hunk 3]
```diff
@@ -16,7 +16,7 @@ @implementation RCTNetworkTask {
   id<RCTURLRequestHandler> _handler;
   dispatch_queue_t _callbackQueue;
   std::mutex _mutex;
-
+  std::atomic<RCTNetworkTaskStatus> _atomicStatus;
```
Comment (hakonk) [inline packages/react-native/Libraries/Network/RCTNetworkTask.mm:19]: The compiler complained about `_status` and `self.status` not having the same types. Thus, I renamed the ivar, but implemented the objective-C property getter. Thoughts @javache ?

File: packages/react-native/Libraries/Network/RCTNetworkTask.h [hunk 4]
```diff
@@ -35,7 +35,7 @@ typedef NS_ENUM(NSInteger, RCTNetworkTaskStatus) {
 @property (nonatomic, copy) RCTURLRequestResponseBlock responseBlock;
 @property (nonatomic, copy) RCTURLRequestProgressBlock uploadProgressBlock;
 
-@property (nonatomic, readonly) RCTNetworkTaskStatus status;
+@property (atomic, readonly) RCTNetworkTaskStatus status;
```
Comment (javache) [inline packages/react-native/Libraries/Network/RCTNetworkTask.h:38]: The atomic attribute doesn't really matter here, since we use a custom getter anyway.

Comment (hakonk) [reply packages/react-native/Libraries/Network/RCTNetworkTask.h:38]: I see. However, as for the API surface of the class, won't it make sense to convey it is indeed atomic even though it is making use of other mechanisms than objc's synchronize?

File: packages/react-native/Libraries/Network/RCTNetworkTask.mm [hunk 5]
```diff
@@ -75,18 +80,18 @@ - (void)start
...
 
 - (void)cancel
 {
-  if (_status == RCTNetworkTaskFinished) {
+  if (_atomicStatus == RCTNetworkTaskFinished) {
     return;
   }
 
-  _status = RCTNetworkTaskFinished;
+  _atomicStatus = RCTNetworkTaskFinished;
```
[the remark this answers is not in the document]
Comment (hakonk) [reply packages/react-native/Libraries/Network/RCTNetworkTask.mm:94]: Good catch!

[the remark this answers is not in the document]
Comment (hakonk) [reply packages/react-native/Libraries/Network/RCTNetworkTask.mm]: https://github.com/facebook/react-native/commit/f0fb515805126ebaa677c80b86bba001079f8ba7

Please see the changes, @javache

**Pull request closed without merging**